examples/gno.land/p/demo/ufmt/ufmt.gno

```diff
@@ -1,299 +1,489 @@
+//	    Currently supports only rune, int.
+//	%q: Formats a string value as a quoted string.
+//	%T: Formats the type of the value.
+//	%v: Formats the value with a default representation appropriate for the value's type
+//	    - nil: <nil>
+//	    - bool: true/false
+//	    - integers: base 10
+//	    - float64: %g format
+//	    - string: verbatim
+//	    - types with String()/Error(): method result
+//	    - others: (unhandled)
+//	%%: Outputs a literal %. Does not consume an argument.
 //
-//		%s: places a string value directly.
-//		    If the value implements the interface interface{ String() string },
-//		    the String() method is called to retrieve the value. Same about Error()
-//		    string.
-//		%c: formats the character represented by Unicode code point
-//		%d: formats an integer value using package "strconv".
-//		    Currently supports only uint, uint64, int, int64.
-//		%f: formats a float value, with a default precision of 6.
-//		%e: formats a float with scientific notation; 1.23456e+78
-//		%E: formats a float with scientific notation; 1.23456E+78
-//		%F: The same as %f
-//		%g: formats a float value with %e for large exponents, and %f with full precision for smaller numbers
-//		%G: formats a float value with %G for large exponents, and %F with full precision for smaller numbers
-//		%t: formats a boolean value to "true" or "false".
-//		%x: formats an integer value as a hexadecimal string.
-//		    Currently supports only uint8, []uint8, [32]uint8.
-//		%c: formats a rune value as a string.
-//		    Currently supports only rune, int.
-//		%q: formats a string value as a quoted string.
-//		%T: formats the type of the value.
-//	     %v: formats the value with a default representation appropriate for the value's type
-//		%%: outputs a literal %. Does not consume an argument.
-func Sprintf(format string, args ...any) string {
-	// we use runes to handle multi-byte characters
+// Unsupported verbs or type mismatches produce error strings like "%!d(string=foo)".
+func Sprintf(format string, a ...any) string {
+	p := newPrinter()
+	p.doPrintf(format, a)
+	return string(p.buf)
+}
+
+// doPrintf parses the format string and writes formatted arguments to the buffer.
+func (p *printer) doPrintf(format string, args []any) {
 	sTor := []rune(format)
 	end := len(sTor)
 	argNum := 0
 	argLen := len(args)
-	buf := ""
 
 	for i := 0; i < end; {
 		isLast := i == end-1
-		c := string(sTor[i])
+		c := sTor[i]
 
-		if isLast || c != "%" {
+		if isLast || c != '%' {
 			// we don't check for invalid format like a one ending with "%"
-			buf += string(c)
+			p.buf.writeRune(c)
 			i++
 			continue
 		}
 
-		verb := string(sTor[i+1])
-		if verb == "%" {
-			buf += "%"
+		verb := sTor[i+1]
+		if verb == '%' {
+			p.buf.writeRune('%')
 			i += 2
 			continue
 		}
 
-		if argNum > argLen {
-			panic("invalid number of arguments to ufmt.Sprintf")
+		if argNum >= argLen {
+			panic("ufmt: not enough arguments")
```

Comment:
Am I crazy, or was the original condition wrong?

Reply:
Not crazy :)